File: LCOF/59-II.dui-lie-de-zui-da-zhi-lcof/dui-lie-de-zui-da-zhi-lcof.py
import queue
class MaxQueue:

    def __init__(self):
        self.deque = queue.deque()  #创建双向队列

    def max_value(self) -> int:
        return max(self.deque) if self.deque else -1

    def push_back(self, value: int) -> None:
        self.deque.append(value)  #往双向队列里塞入value

    def pop_front(self) -> int:
        return self.deque.popleft() if self.deque else -1 #queue.deque.popleft 获取最左边一个元素，并在队列中删除


# Your MaxQueue object will be instantiated and called as such:
# obj = MaxQueue()
# param_1 = obj.max_value()
# obj.push_back(value)
# param_3 = obj.pop_front()

File: LCOF/59-II.dui-lie-de-zui-da-zhi-lcof/dui-lie-de-zui-da-zhi-lcof.go
/*
双队列

    入队
        q1 : 单向队列，用于存储进队的值
        max ：双向队列，较大值进队，较小值出队，用于存储最大值，呈递减序列
    出队
        q1 : 首部出队
        max：队列的首部与 q1 出队的值作比较，如果相同，那么一起出队
    若队列为空，pop_front 和 max_value 返回 -1
*/

type MaxQueue struct {
    q1 []int
    max []int
}


func Constructor() MaxQueue {
    return MaxQueue{
        make([]int,0),
        make([]int,0),
    }
}


func (this *MaxQueue) Max_value() int {
    if len(this.max) == 0{
        return -1
    }
    return this.max[0]
}


func (this *MaxQueue) Push_back(value int)  {
    this.q1 = append(this.q1,value)  //将value入q1队列
    for len(this.max) != 0 && value > this.max[len(this.max)-1]{
        this.max = this.max[:len(this.max)-1]
    }
    this.max = append(this.max,value)
}


func (this *MaxQueue) Pop_front() int {
    n := -1
    if len(this.q1) != 0{
        n = this.q1[0]
        this.q1 = this.q1[1:]
        if this.max[0] == n{
            this.max = this.max[1:]
        }
    }
    return n
}


/**
 * Your MaxQueue object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Max_value();
 * obj.Push_back(value);
 * param_3 := obj.Pop_front();
 */
